src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs

```diff
@@ -781,43 +781,49 @@ protected async Task<ImmutableArray<Diagnostic>> GetSortedDiagnosticsAsync(Solut
                 var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
                 var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers, GetAnalyzerOptions(project), cancellationToken);
                 var allDiagnostics = await compilationWithAnalyzers.GetAllDiagnosticsAsync().ConfigureAwait(false);
-                diagnostics.AddRange(allDiagnostics.Where(IsIncludedDiagnostic));
+
+                diagnostics.AddRange(allDiagnostics.Where(diagnostic => IsCompilerDiagnosticIncluded(diagnostic, compilerDiagnostics)));
             }
 
             var results = SortDistinctDiagnostics(diagnostics);
             return results.ToImmutableArray();
+        }
 
-            // Local function
-            bool IsIncludedDiagnostic(Diagnostic diagnostic)
-            {
-                return !IsCompilerDiagnostic(diagnostic)
-                    || IsCompilerDiagnosticIncluded(diagnostic);
-            }
+        /// <summary>
+        /// Extension point to allow inheriting classes to include diagnostics which would otherwise be filtered. The default implementation returns false
+        /// </summary>
+        /// <param name="diagnostic">the diagnostic which is subject to filter processes.</param>
+        /// <param name="compilerDiagnostics">level of diagnostic used to filter compiler diagnostics</param>
+        /// <returns>return true to exclude a diagnostic, false to leave it up to internal logic.</returns>
+        protected virtual bool IsCompilerDiagnosticIncluded(Diagnostic diagnostic, CompilerDiagnostics compilerDiagnostics)
+        {
+            return !IsCompilerDiagnostic(diagnostic)
```

Comment:
📝 The `!IsCompilerDiagnostic(diagnostic)` should still be up in the caller. When `IsCompilerDiagnosticIncluded` is called, we expect the argument to always (already) be a compiler diagnostic.

Reply:
Mhm, i suppose i changed that after you started the review. At least without exclamation mark tests failed and i restored the code to the change you mention here.